walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -114,4 +145,56 @@
 
     return possibleSlots;
   }
+
+  /**
+   * Find the intersection of two collections of slots. The intersection between two individual
+   * slots is the slot that is contained within the other one. If none of the two slots is contained
+   * within the other one, then their intersection is empty.
+   *
+   * @param mandatory A collection of {@code TimeRange}s that represents meeting slots for mandatory
+   *     attendees
+   * @param optional A collection of {@code TimeRange}s that represents meeting slots for optional
+   *     attendees
+   * @return The intersection of these collections as specified above.
+   */
+  public Collection<TimeRange> intersection(
+      Collection<TimeRange> mandatory, Collection<TimeRange> optional) {
+    Iterator<TimeRange> itMandatory = mandatory.iterator();
+    Iterator<TimeRange> itOptional = optional.iterator();
+
+    TimeRange currentMandatory = itMandatory.next();
+    TimeRange currentOptional = itOptional.next();
+
+    List<TimeRange> intersection = new ArrayList<TimeRange>();
+
+    /* it is guaranteed that one of the slot collections will reach its final slot and
+    then stay there until the other collection catches up */
+    while (itMandatory.hasNext() || itOptional.hasNext()) {
+
+      /* if one slot completely contains the other, keep the smaller slot */
+      if (currentMandatory.contains(currentOptional)) {
+        intersection.add(currentOptional);
+      } else if (currentOptional.contains(currentMandatory)) {
+        intersection.add(currentMandatory);
+      }
+
+      /* point to whichever slot comes earlier (this is why it is guaranteed that
+      once one collection reaches its end slot, getting the next slot
+      will not be attempted until the other slot collection catches up to it */
+      if (currentMandatory.end() > currentOptional.end()) {
+        currentOptional = itOptional.next();
+      } else {
+        currentMandatory = itMandatory.next();
+      }
+    }
+
+    /* in case there are any remaining last slots, check again which one we keep (if any) */
+    if (currentMandatory.contains(currentOptional)) {
+      intersection.add(currentOptional);
+    } else if (currentOptional.contains(currentMandatory)) {
+      intersection.add(currentMandatory);
+    }
```

Comment:
What about the case where neither contains the other, but there exists a large enough overlap?

```
|---mand--|
     |---opt--|
  =  |----|   (assuming duration <= 4.
```

It doesn't look like this code, nor tests, handle this case.